### ISSUE TITLE
[Question] differences between MSE loss and Huber loss

### ISSUE THREAD

**Issue by huangzehao** [nagadomi/waifu2x#159]
Hi, nagadomi. Thanks for sharing this amazing project.
I have some questions about the implementation of waifu2x.

First, have you benchmarked the performance differences between MSE loss and Huber loss?
In my opinion, Huber loss may work better than MSE loss when the input image has slight noise. But when the input image is clear, I think MSE loss is enough.

Second, what's the difference between up_conv7l and up_conv7? I did't find the model definition of up_conv7l in this repo. I guess that up_conv7l is wider than up_conv7.  Am I right?

Third, I have implemented SRResNet in this [paper](https://arxiv.org/abs/1609.04802), including MSE loss and Perceptual loss. I found the performance of MSE loss is pretty good. But in some images, there are overflow problem in my implementation. 
Following is my 4x upscale result of w2nn.png. The edge of stairs in this result is very clear. But there are some strange overflow pixel.. 
Do you have any idea to handle this problem? I implemented SRResNet also in torch.
![outputs](https://cloud.githubusercontent.com/assets/13661914/20957298/ed77545e-bc89-11e6-954e-4369783fc984.png)

Thanks!

**Comment by nagadomi** [comment]
>I think MSE loss is enough.

I think so. The threshold parameter of Huber loss in waifu2x is large. When the target image is clear, L1 loss is used only in the first few epochs, otherwise L2 loss(MSE) is used. But the target image is higher resolution than the input image, typically it's more noisy than the input image. But of course, L1 loss and Huber loss do not maximize PSNR metric directly.

>I did't find the model definition of up_conv7l in this repo.

`upconv_7l` can be found in [dev branch](https://github.com/nagadomi/waifu2x/tree/dev). Also I've added `srresnet_12l`.
```
git checkout -b dev origin/dev
```

> But in some images, there are overflow problem in my implementation.

Maybe you can prevent this problem by clipping the output layer. See [upconv_7](https://github.com/nagadomi/waifu2x/blob/3c28debad9bb048e87164ae398da3e462918f3bc/lib/srcnn.lua#L273), [InplaceClip01](https://github.com/nagadomi/waifu2x/blob/3c28debad9bb048e87164ae398da3e462918f3bc/lib/InplaceClip01.lua) (also you can use `nn.Clamp`). 
But I have never met this problem. If your dataset is not too small, there might be some bugs.

**Comment by huangzehao** [comment]
Thanks for your replay and suggestion!
Dev branch is awesome! You have done so many experiments to increase the performance of SR.
I trained my model by 50000 training samples from imagenet validation set. Maybe there are some bugs in my implementation.
I think I should read your code carefully. Since there are many implementation details in this repo, it is a best example for people who want to develop better SR techniques.
Thanks for your greate work! 
If you need help to do experiments, I can contribute!

**Comment by nagadomi** [comment]
Thanks

**Comment by nagadomi** [comment]
I'm not sure if this is correct. I think that problem is related to batch normlization, if it's not simple coding bug.

**Comment by huangzehao** [comment]
It's really caused by batch normlization. Since the parameters obtained by training set is not suitable for some image. I set model:evaluate() to model:training() and it solved. 
Thanks!
